user/tlu/module/src/TluRawEvent2StdEventConverter.cc

```diff
@@ -29,23 +29,36 @@ bool TluRawEvent2StdEventConverter::Converting(eudaq::EventSPC d1, eudaq::Standa
     d2->SetTag(TLU+stm+"_TRG", std::to_string(d1->GetTriggerN()));
   }
 
-  auto triggersFired = d1->GetTag("TRIGGER" , "NAN");
-  auto finets0 = d1->GetTag("FINE_TS0", "NAN");
-  auto finets1 = d1->GetTag("FINE_TS1", "NAN");
-  auto finets2 = d1->GetTag("FINE_TS2", "NAN");
-  auto finets3 = d1->GetTag("FINE_TS3", "NAN");
-  auto finets4 = d1->GetTag("FINE_TS4", "NAN");
-  auto finets5 = d1->GetTag("FINE_TS5", "NAN");
+  auto triggersFired = std::stoi(d1->GetTag("TRIGGER" , "NAN"), nullptr, 2); // interpret as binary
+  auto finets0 = std::stoi(d1->GetTag("FINE_TS0", "NAN"));
+  auto finets1 = std::stoi(d1->GetTag("FINE_TS0", "NAN"));
+  auto finets2 = std::stoi(d1->GetTag("FINE_TS0", "NAN"));
+  auto finets3 = std::stoi(d1->GetTag("FINE_TS0", "NAN"));
+  auto finets4 = std::stoi(d1->GetTag("FINE_TS0", "NAN"));
+  auto finets5 = std::stoi(d1->GetTag("FINE_TS0", "NAN"));
 
   // Set times for StdEvent in picoseconds (timestamps provided in nanoseconds):
-  auto ts = (d1->GetTimestampBegin() * 1000) & 0xFFFFFFFFFFFFFF00 + static_cast<uint64_t>(25. / 32. * std::stoi(finets0));
+  // Note: __builtin_popcount counts the "ones" in a binary, see here:
+  // https://www.geeksforgeeks.org/count-set-bits-in-an-integer/
+  // I.e. add up all timestamps for which the triggersFired bit is 1 and divide by number of active triggers:
+  // Factor 25./32. -> fine timestamp in 781ps binning
+  auto finets = ((finets0 * ((triggersFired >> 0) & 0x1))
+               + (finets1 * ((triggersFired >> 1) & 0x1))
+               + (finets2 * ((triggersFired >> 2) & 0x1))
+               + (finets3 * ((triggersFired >> 3) & 0x1))
+               + (finets4 * ((triggersFired >> 4) & 0x1))
+               + (finets5 * ((triggersFired >> 5) & 0x1)))
+               / __builtin_popcount(triggersFired); // count "ones" in binary
 
-  d2->SetTimeBegin(d1->GetTimestampBegin() * 1000);
+  auto ts = (d1->GetTimestampBegin() * 1000) & 0xFFFFFFFFFFFFFF00 + static_cast<uint64_t>(25. / 32. * finets);
+
+  d2->SetTimeBegin(ts);
   d2->SetTimeEnd(d1->GetTimestampEnd() * 1000);
-  d2->SetTimestamp(d1->GetTimestampBegin(), d1->GetTimestampEnd(), d1->IsFlagTimestamp());
+  d2->SetTimestamp(ts, d1->GetTimestampEnd(), d1->IsFlagTimestamp());
```

Comment:
How do we treat the end? right now it might well be before `start` because we add sth to `start` but then use it as end.

Reply:
In the producer it's set to eventStart + 25ns.
So should we also just do evenEnd + fine and shift it by the same amount?

Reply:
ah, okay. Hm, not necessary I guess. I was thinking whether it would make sense to actually use the coincidence window, but maybe even that doesn't make too much sense.

Reply:
The 25ns is just an arbitrary number without any meaning, so the coincidence window might not be a bad idea. I see how that's feasible.

Anyway, we're not using this value for anything in the analysis (especially now that we have the fine TS, the coincidence time somewhat looses its meaning).